File: heap/05_find_median_from_data_stream.py
"""
PROBLEM: Find Median from Data Stream (LeetCode 295)
DIFFICULTY: Hard
PATTERN: Heap - Two Heaps (max-heap for lower half, min-heap for upper half)

DESCRIPTION:
Design a data structure that supports adding numbers and finding median at any time.

EXAMPLE:
add(1), add(2) -> median = 1.5
add(3) -> median = 2.0
add(4), add(5) -> median = 3.0

TIME COMPLEXITY: O(log n) add, O(1) findMedian
SPACE COMPLEXITY: O(n)
"""

import heapq


class MedianFinder:
    def __init__(self) -> None:
        # max_heap: lower half (negate to simulate max-heap)
        self.max_heap: list = []
        # min_heap: upper half
        self.min_heap: list = []

    def addNum(self, num: int) -> None:
        # Push to max_heap first
        heapq.heappush(self.max_heap, -num)
        # Balance: ensure max_heap top <= min_heap top
        if self.min_heap and (-self.max_heap[0]) > self.min_heap[0]:
            heapq.heappush(self.min_heap, -heapq.heappop(self.max_heap))
        # Balance sizes: max_heap can have at most 1 extra
        if len(self.max_heap) > len(self.min_heap) + 1:
            heapq.heappush(self.min_heap, -heapq.heappop(self.max_heap))
        elif len(self.min_heap) > len(self.max_heap):
            heapq.heappush(self.max_heap, -heapq.heappop(self.min_heap))

    def findMedian(self) -> float:
        if len(self.max_heap) > len(self.min_heap):
            return float(-self.max_heap[0])
        return (-self.max_heap[0] + self.min_heap[0]) / 2.0


if __name__ == "__main__":
    mf = MedianFinder()
    ops = [1, 2, 3, 4, 5]
    for num in ops:
        mf.addNum(num)
        print(f"After adding {num}: median = {mf.findMedian()}")

    mf2 = MedianFinder()
    mf2.addNum(1)
    mf2.addNum(2)
    print(f"Median after [1,2]: {mf2.findMedian()}")
    mf2.addNum(3)
    print(f"Median after [1,2,3]: {mf2.findMedian()}")

File: heap/05_find_median_from_data_stream.cpp
#include <iostream>
#include <queue>
using namespace std;

/*
Problem Description:
Design a data structure that supports adding numbers and finding median at any time.

Sample Input/Output:
1) add 1,2 -> median = 1.5
2) add 3 -> median = 2
3) add 4,5 -> median = 3

Code Explanation:
Use two heaps:
- maxHeap for lower half
- minHeap for upper half
Balance sizes so maxHeap has equal or one extra element.
*/

class MedianFinder {
private:
    priority_queue<int> maxHeap;
    priority_queue<int, vector<int>, greater<int>> minHeap;

public:
    void addNum(int num) {
        maxHeap.push(num);
        minHeap.push(maxHeap.top());
        maxHeap.pop();

        if (minHeap.size() > maxHeap.size()) {
            maxHeap.push(minHeap.top());
            minHeap.pop();
        }
    }

    double findMedian() {
        if (maxHeap.size() == minHeap.size()) {
            return (maxHeap.top() + minHeap.top()) / 2.0;
        }
        return maxHeap.top();
    }
};

int main() {
    MedianFinder mf;

    mf.addNum(1); mf.addNum(2);
    cout << "Output 1: " << mf.findMedian() << "\n";

    mf.addNum(3);
    cout << "Output 2: " << mf.findMedian() << "\n";

    mf.addNum(4); mf.addNum(5);
    cout << "Output 3: " << mf.findMedian() << "\n";
    return 0;
}
